app/models/models.py

```diff
@@ -0,0 +1,151 @@
+from datetime import datetime, timedelta
+from decimal import Decimal
+from math import floor
+from typing import Any, Dict, List, Set
+
+from config import (
+	DATE_TIME_FORMAT,
+	FREE_MINUTES,
+	TIME_FRAME_HOURS,
+	TIME_FRAME_TARIFFS
+)
+
+class ParkingLotFullError(Exception):
+	"""Custom error that is raised when the Parking Lot is full"""
+	pass
+
+class ParkingLotLocationEmptyError(Exception):
+	"""Custom error that is raised when the Parking Lot location is empty"""
+	pass
+
+class Ticket:
+	def __init__(self, license_plate: str, tariff: str, location: int) -> None:
+		"""Ticket constructor
+
+		Args:
+			license_plate (str): Car license plate.
+			tariff (str): Tariff for the ticket.
+			location (int): Parking Lot location.
+
+		Returns:
+			None
+		"""
+		self.car: str = license_plate
+		self.tariff: str = tariff
+		self.location: int = location
+		self.start: str = datetime.now().strftime(DATE_TIME_FORMAT)
+		self.finish: str = ""
+		self.fee: str = ""
+
+	def _get_total_time(self) -> timedelta:
+		"""Get the total time for the ticket
+
+		Returns:
+			timedelta: Total time for the ticket.
+		"""
+		try:
+			start = datetime.strptime(self.start, DATE_TIME_FORMAT)
+			finish = datetime.strptime(self.finish, DATE_TIME_FORMAT)
+		except ValueError as e:
+			raise ValueError("Invalid date format") from e
+		return finish - start
+
+	def calculate_fee(self, end_datetime: datetime = datetime.now()) -> None:
+		"""Calculate the fee for the ticket
+
+		Returns:
+			None
+		"""
+		self.finish = end_datetime.strftime(DATE_TIME_FORMAT)
+		delta = self._get_total_time()
+		total_minutes = delta / timedelta(minutes=1)
+
+		if total_minutes <= FREE_MINUTES:
+			self.fee = "0.00"
+			return
+
+		time_frame = TIME_FRAME_HOURS.get(self.tariff)
+		tariff = TIME_FRAME_TARIFFS.get(self.tariff)
+
+		total_time = delta / timedelta(hours=time_frame)
+
+		if not total_time.is_integer():
+			total_time += 1
+
+		self.fee = str(Decimal(tariff) * Decimal(floor(total_time)))
+
+
+class Car:
+	def __init__(self, license_plate: str) -> None:
+		"""Car constructor.
+
+		Args:
+			license_plate (str): Car license plate.
+
+		Returns:
+			None
+		"""
+		self.license_plate = license_plate
+
+
+class ParkingLot:
+	def __init__(self, total_spots: int) -> None:
+		"""Parking Lot constructor.
+
+		Returns:
+			None
+		"""
+		self.total_spots: int = total_spots
+		self.available_spots: Set[int] = set(range(1, self.total_spots+1))
+		self.occupied_spots: Dict[int, Ticket] = {}
+
+	def add_car(self, car: Car, tariff: str) -> Ticket:
+		"""Add car to the Parking Lot.
+
+		Returns:
+			Ticket: Ticket object with the car information.
+
+		Raises:
+			ParkingLotFullError: If the Parking Lot is full.
+		"""
+
+		try:
+			available_spot = self.available_spots.pop()
+		except KeyError as e: # Full parking Lot
+			raise ParkingLotFullError("No free space") from e
```

Comment:
It is non-trivial that KeyError means no free space. Is it possible to improve this?

Reply:
Yes. I improved it by doing it more "pythonic":

```python
available_spot = len(self.occupied_spots) + 1
if available_spot <= self.total_spots:
	base_tariff_cost = self.time_frame_tariffs.get(tariff)
	self.occupied_spots[available_spot] = Ticket(car.license_plate, tariff, available_spot, base_tariff_cost)
	return self.occupied_spots[available_spot]
raise ParkingLotFullError("No free space")
```

For this case, I made an improvement in my Parking Lot logic. I removed the set that I was using to get the available spots. Instead, I use the dictionary length + 1 to get the next spot to use if it is <= to TOTAL SPOTS.